### PR TITLE
Updated PHPStan & related fixes

### PR DESCRIPTION
**Pull request by ondrejmirtes**
I removed PHPStan from `allow_failures` because I think it should pass now. Let's see if the build passes :)

### CODE REVIEW

**Comment by antograssiot** [comment]
Out of curiosity, is there a real advantage in running it against php 7.0 and 7.1 ?
And if there is what about php 5.6 ?

**Comment by ondrejmirtes** [comment]
@antograssiot There can be some differences in language functions between 7.0 and 7.1. And I run PHPStan only on 7.x because it supports >= 7.0 only :)

File: src/Database/Statement/BufferedStatement.php [hunk 1]
```diff
@@ -91,7 +91,6 @@ public function fetch($type = 'num')
             return $row;
         }
 
-        $this->_fetchType = $type;
         $record = parent::fetch($type);
```
Comment (markstory) [inline src/Database/Statement/BufferedStatement.php]: Does _fetchType get set in the parent?

File: src/Shell/I18nShell.php [hunk 2]
```diff
@@ -87,7 +87,7 @@ public function init($language = null)
             $language = $this->in('Please specify language code, e.g. `en`, `eng`, `en_US` etc.');
         }
         if (strlen($language) < 2) {
-            return $this->abort('Invalid language code. Valid is `en`, `eng`, `en_US` etc.');
+            $this->abort('Invalid language code. Valid is `en`, `eng`, `en_US` etc.');
```
Comment (markstory) [inline src/Shell/I18nShell.php]: The return is helpful when abort() is mocked.

File: src/Database/Statement/BufferedStatement.php (hunk 1 above)
Comment (ADmad) [inline src/Database/Statement/BufferedStatement.php]: I did a search in the codebase and there is no there reference to `_fetchType`.

**Comment by ADmad** [comment]
Build is failing https://travis-ci.org/cakephp/cakephp/builds/192196094

Coincidentally I was working on this as well https://github.com/cakephp/cakephp/compare/3.next...static-analysis  Good to see the new phpstan version requires lesser ignores.

File: src/Shell/I18nShell.php (hunk 2 above)
Comment (ADmad) [inline src/Shell/I18nShell.php]: I guess moving the `return` to next line should make phpstan happy.

Comment (dereuromark) [inline src/Shell/I18nShell.php]: Yes, a newline would do the trick.
Also, isn't it now just an exception being thrown? That should be easier to work with now in general, I guess.

**Comment by ondrejmirtes** [comment]
I noticed that the build is failing, because of different reasons in each PHP version :) I'll look at it soon.

**Comment by ADmad** [comment]
The declaration of `$_connection` property in `Cake\Database\Driver` seems to be the cause of failures.

**Comment by ADmad** [comment]
@ondrejmirtes Ping

**Comment by ondrejmirtes** [comment]
Don't worry, I still have this PR in the back of my mind, I'll finish it
once I have a few moments for to do it 😊

File: src/Database/Statement/BufferedStatement.php (hunk 1 above)
Comment (markstory) [inline src/Database/Statement/BufferedStatement.php]: Ok.

File: src/Shell/I18nShell.php (hunk 2 above)
Comment (dereuromark) [inline src/Shell/I18nShell.php]: @markstory with the exception here it doesn't need a return anymore now, so the code is OK, right?

**Comment by ondrejmirtes** [comment]
I updated the PR so let's see if it builds. There are two weird code samples remaining:

ConnectionManager::setConfig
Log::setConfig

Both call and return `static::_setConfig` which is from StaticConfigTrait and marked as`@return void` but these methods read its result and return it. Is there anything I don't see? Which side is wrong?

**Comment by ondrejmirtes** [comment]
The only failing code lines are the ones I mentioned in previous comment. ^^

**Comment by dereuromark** [comment]
@ondrejmirtes You are right, something is not quite right here, I opened a PR for discussion around it.

**Comment by ADmad** [comment]
@ondrejmirtes Can you please rebase your phpstan branch on top of upstream master.

**Comment by ondrejmirtes** [comment]
@ADmad I think you meant on top of `3.next`? Because there's no signs of PHPStan in master. So I rebased it on `3.next`, awaiting further instructions :)

**Comment by ADmad** [comment]
Yes sorry, I meant `3.next` not `master`. I see that rhe phpstan job now passes on php 7 but not on php 7.1. So once that's sorted out this is good to merge.

**Comment by markstory** [comment]
PHPStan build is [failing](https://travis-ci.org/cakephp/cakephp/jobs/197760632)

**Comment by ondrejmirtes** [comment]
@ADmad @markstory The problem is that APCu is missing from Travis CI on 7.1. So I'm trying to install it in `before_script`, let's see if it works.

**Comment by ondrejmirtes** [comment]
The build currently fails because PHPUnit 6.0 contains some BC breaks. Please let me know once this is solved in upstream do I can rebase and verify that the build passes.

**Comment by ADmad** [comment]
@ondrejmirtes phpunit issue is fixed. Please rebase.

**Comment by ondrejmirtes** [comment]
Rebased. Let's see if it builds :)

File: .travis.yml [hunk 3]
```diff
@@ -90,14 +86,15 @@ before_script:
   - composer install --prefer-dist --no-interaction
   - if [[ $PHPCS = 3 ]]; then composer remove --dev cakephp/cakephp-codesniffer; fi
   - if [[ $PHPCS = 3 ]]; then composer require --dev squizlabs/php_codesniffer='3.0.*@dev' cakephp/cakephp-codesniffer='3.0.*@dev'; fi
+  - printf "yes\n" | pecl install apcu
```
Comment (markstory) [inline .travis.yml]: This is causing some of the builds to fail.

Comment (ondrejmirtes) [inline .travis.yml]: I haven't noticed that .travis.yml already contains installation of APCu elsewhere so I changed the commit: https://github.com/cakephp/cakephp/pull/10044/commits/b8c60da7957aaf835a24f9fb83bcdab5bae16890

**Comment by ondrejmirtes** [comment]
The build passes :)

**Comment by markstory** [comment]
Thanks!

**Pull request merged**